### ISSUE TITLE
NEGemm: matrix sizes and orders

### ISSUE THREAD

**Issue by gary-wjc** [ARM-software/ComputeLibrary#862]
I have a few questions about the usage of Neon matrix multiply for the case of C = AB.

(1). The order of source matrices:
I know from the example that, when all the matrices are row-major, the use is straightforward: initialize tensor A, B and C in default layout, then configure and run NEGemm. But what is the most efficient way to do gemm only once when one or two of the source matrices are column-major ?

(2). The scheduler
There're 2 scheduler (CPP & OpenMP) available. Currently which one is more efficient in most cases ?

(3). Flat matrices
Are there any APIs specially designed for GEMM with very small M or N using Neon other than the general API in NEGemm ? 

Thanks.

**Comment by morgolock** [comment]
Hi @gary-wjc 

1) Why not reshaping all the matrices involved in the computation to be row-major? 

2) Difficult to answer, I'd say there shouldn't be big performance differences between the two schedulers. You should assess this on a specific configuration: hardware, OS, build toolchain and choose the one that suits your needs better.

3) Could you please tell us an example of the values MNK?

Hope this helps.

**Comment by gary-wjc** [comment]
Thank you for your answer @morgolock  ~

1. In some cases one of the source matrices can be column-major. Take the widely-used transformer model in NLP for example. Most of the GEMM operations envolved in the inference stage (like the generation of q, k and v matrix in attention layer, and the operations in FFN layer) can be configured as all-row-major GEMM, but the dot operation of q and k to generate logits must have one input matrix in column-major form.

2. An example of special MNK size that would occur in the inference process when using transformer model: M = 3, N = 1280 and K = 400, for the first GEMM in FFN layer, when beam-width = 3, hidden-size = 400 and filter-size = 1280.

**Comment by morgolock** [comment]
Hi @gary-wjc 

Thanks for the additional details.

1. NEGEMM assumes all the operands have the same ordering, you cannot mix column-major with row-major operands. Unfortunately you would have to transform any column-major operand to row-major before callling NEGEMM::configure() and NEGEMM::run().

2. I'd suggest you do experiment with NEGEMM. This function can choose among multiple highly optimized assembly kernels based on input shapes and underlying hardware support. Some of these kernels are optimized to deal with typical input shapes from workloads in mainstream models.

Hope this helps.
